File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/client/OneDriveProvider.java
package ru.kmorozov.onedrive.client;

import ru.kmorozov.onedrive.client.resources.Drive;
import ru.kmorozov.onedrive.client.authoriser.AuthorisationProvider;
import ru.kmorozov.onedrive.client.downloader.ResumableDownloaderProgressListener;

import java.io.File;
import java.io.IOException;
import java.util.Date;

public interface OneDriveProvider {

    // Read only operations

    Drive getDefaultDrive() throws IOException;

    OneDriveItem getRoot() throws IOException;

    OneDriveItem[] getChildren(OneDriveItem parent) throws IOException;

    OneDriveItem[] getChildren(String id) throws IOException;

    OneDriveItem getPath(String path) throws IOException;

    OneDriveItem getItem(String id) throws IOException;

    // Write operations

    OneDriveItem replaceFile(OneDriveItem parent, File file) throws IOException;

    OneDriveItem uploadFile(OneDriveItem parent, File file) throws IOException;

    OneDriveUploadSession startUploadSession(OneDriveItem parent, File file) throws IOException;

    void uploadChunk(OneDriveUploadSession session) throws IOException;

    OneDriveItem updateFile(OneDriveItem item, Date createdDate, Date modifiedDate) throws IOException;

    OneDriveItem createFolder(OneDriveItem parent, String name) throws IOException;

    void download(OneDriveItem item, File target, ResumableDownloaderProgressListener progressListener, int chunkSize) throws IOException;

    void download(OneDriveItem item, File target, ResumableDownloaderProgressListener progressListener) throws IOException;

    void delete(OneDriveItem remoteFile) throws IOException;

    OneDriveItem[] search(String query) throws IOException;

    class FACTORY {

        public static OneDriveProvider readOnlyApi(final AuthorisationProvider authoriser) {
            return new ROOneDriveProvider(authoriser);
        }

        public static OneDriveProvider readWriteApi(final AuthorisationProvider authoriser) {
            return new RWOneDriveProvider(authoriser);
        }
    }
}

File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/client/OneDriveProvider.kt
package ru.kmorozov.onedrive.client

import ru.kmorozov.onedrive.client.resources.Drive
import ru.kmorozov.onedrive.client.authoriser.AuthorisationProvider
import ru.kmorozov.onedrive.client.downloader.ResumableDownloaderProgressListener

import java.io.File
import java.io.IOException
import java.util.Date

interface OneDriveProvider {

    // Read only operations

    val defaultDrive: Drive

    val root: OneDriveItem?

    @Throws(IOException::class)
    fun getChildren(parent: OneDriveItem): Array<OneDriveItem>

    @Throws(IOException::class)
    fun getChildren(id: String): Array<OneDriveItem>

    @Throws(IOException::class)
    fun getPath(path: String): OneDriveItem

    @Throws(IOException::class)
    fun getItem(id: String): OneDriveItem

    // Write operations

    @Throws(IOException::class)
    fun replaceFile(parent: OneDriveItem, file: File): OneDriveItem

    @Throws(IOException::class)
    fun uploadFile(parent: OneDriveItem, file: File): OneDriveItem

    @Throws(IOException::class)
    fun startUploadSession(parent: OneDriveItem, file: File): OneDriveUploadSession

    @Throws(IOException::class)
    fun uploadChunk(session: OneDriveUploadSession)

    @Throws(IOException::class)
    fun updateFile(item: OneDriveItem, createdDate: Date, modifiedDate: Date): OneDriveItem

    @Throws(IOException::class)
    fun createFolder(parent: OneDriveItem, name: String): OneDriveItem

    @Throws(IOException::class)
    fun download(item: OneDriveItem, target: File, progressListener: ResumableDownloaderProgressListener, chunkSize: Int)

    @Throws(IOException::class)
    fun download(item: OneDriveItem, target: File, progressListener: ResumableDownloaderProgressListener)

    @Throws(IOException::class)
    fun delete(remoteFile: OneDriveItem)

    @Throws(IOException::class)
    fun search(query: String): Array<OneDriveItem>

    object FACTORY {

        fun readOnlyApi(authoriser: AuthorisationProvider): OneDriveProvider {
            return ROOneDriveProvider(authoriser)
        }

        fun readWriteApi(authoriser: AuthorisationProvider): OneDriveProvider {
            return RWOneDriveProvider(authoriser)
        }
    }
}
